sourcing-new/src/main/scala/ch/epfl/bluebrain/nexus/sourcingnew/projections/Projection.scala

```diff
@@ -0,0 +1,88 @@
+package ch.epfl.bluebrain.nexus.sourcingnew.projections
+
+import java.io.{PrintWriter, StringWriter}
+
+import akka.persistence.query.Offset
+import cats.effect.{Async, ContextShift, IO, LiftIO}
+import cats.implicits._
+import fs2.Stream
+import io.circe.Decoder
+import io.circe.parser.decode
+
+import scala.concurrent.Future
+
+/**
+  * A Projection represents the process to transforming an event stream into a format that's efficient for consumption.
+  * In terms of CQRS, the events represents the format in which data is written to the primary store (the write
+  * model) while the result of a projection represents the data in a consumable format (the read model).
+  *
+  * Projections replay an event stream
+  */
+trait Projection[F[_], A] {
+
+  /**
+    * Records progress against a projection identifier.
+    *
+    * @param id       the projection identifier
+    * @param progress the offset to record
+    * @return a future () value
+    */
+  def recordProgress(id: ProjectionId, progress: ProjectionProgress): F[Unit]
+
+  /**
+    * Retrieves the progress for the specified projection projectionId. If there is no record of progress
+    * the [[ProjectionProgress.NoProgress]] is returned.
+    *
+    * @param id an unique projectionId for a projection
+    * @return a future progress value for the specified projection projectionId
+    */
+  def progress(id: ProjectionId): F[ProjectionProgress]
+
+  /**
+    * Record a specific event against a index failures log projectionId.
+    *
+    * @param id             the project identifier
+    * @param failureMessage the failure message to persist
+    */
+  def recordFailure(id: ProjectionId, failureMessage: FailureMessage[A], f: Throwable => String = Projection.stackTraceAsString): F[Unit]
+
+  /**
+    * An event stream for all failures recorded for a projection.
+    *
+    * @param id the projection identifier
+    * @return a source of the failed events
+    */
+  def failures(id: ProjectionId):Stream[F, (A, Offset, String)]
+}
+
+object Projection {
+
+  private [projections] def wrapFuture[F[_]: LiftIO, A](f: => Future[A])
+                                                       (implicit cs: ContextShift[IO]): F[A] =
+    IO.fromFuture(IO(f)).to[F]
+
+  private [projections] def decodeOption[A: Decoder, F[_]]
+                                  (input: Option[String], defaultValue: A)
+                                  (implicit F: Async[F]): F[A] =
+    input match {
+      case Some(value) => F.fromTry(decode[A](value).toTry)
+      case None => F.pure(defaultValue)
+    }
+
+  private [projections] def decodeError[A: Decoder, B: Decoder](input: (String, String, String)): Option[(A, B, String)] = {
+    (decode[A](input._1), decode[B](input._2), Right(input._3)).tupled.toOption
+    }
+
+  /**
+    * Allows to get a readable stacktrace
+    * @param t
+    * @return
+    */
+  def stackTraceAsString(t: Throwable) = {
```

Comment:
is this used anywhere?

Reply:
It is used to format stacktraces in a human readable way in one of the colums in projection failures.